### PR TITLE
feat: use PolkaVM disassembler

### PR DESCRIPTION
**Pull request by 0xf333**
## Desciption
  - Integratd the disassembler
  - updated `asm.test` to check for the correct patterns in the assembly text.
  - addressed tsconfig conflict

### Usage example 
```bash
zksolc contract.sol --asm --output-dir './tmpdir'
```
The assembly `.zasm` file will be outputed in the `tmpdir` directory

---

### `--asm` flag test
```bash
make test-cli
```
Now `asm.test` working properly

### CODE REVIEW

**Comment by 0xf333** [comment]
### Introduced change
  - Integrated the disassembler
  - updated `.gitginore` 

### Usage example 
```bash
zksolc contract.sol --asm --output-dir './tmpdir'
```
The assembly `.zasm` file will be outputed in the `tmpdir` directory

**Comment by 0xf333** [comment]
### Introduced change
   - Now asm print in `main_inner` to output the correct assembly text.
   - Now `asm.test` to check for the correct patterns in the assembly text.
   
### Usage
```bash
make test-cli
```
Now `asm.test` working properly

**Comment by 0xf333** [comment]
@xermicus

File: crates/solidity/src/zksolc/main.rs [hunk 1]
```diff
@@ -190,18 +194,36 @@ fn main_inner() -> anyhow::Result<()> {
...
-                println!(
-                    "Contract `{}` assembly:\n\n{}",
-                    path, contract.build.assembly_text
-                );
+                let program_blob = match ProgramBlob::parse(bytescode.as_slice()) {
+                    Ok(blob) => blob,
+                    Err(error) => {
+                        bail!("Failed to parse program blob: {}", error);
+                    }
+                };
```
Comment (xermicus) [inline crates/solidity/src/zksolc/main.rs:205]: Total nitpick. But you can do instead:

```suggestion
                let program_blob = ProgramBlob::parse(bytescode.as_slice())
                    .map_err(|error| anyhow::anyhow!("Failed to parse program blob: {error}"))?;
```

More concise and I personally find it easier to read. But it's a question of style and personal preference so I'll leave it up to you.

File: crates/solidity/src/zksolc/main.rs [hunk 2]
```diff
@@ -190,18 +194,36 @@ fn main_inner() -> anyhow::Result<()> {
...
+
+                let disassembler_object =
```
Comment (xermicus) [inline crates/solidity/src/zksolc/main.rs:207]: Another nitpick, in rust land we would name this variable just `disassembler` :upside_down_face:

File: crates/solidity/src/zksolc/main.rs [hunk 3]
```diff
@@ -190,18 +194,36 @@ fn main_inner() -> anyhow::Result<()> {
...
+                    Disassembler::new(&program_blob, DisassemblyFormat::Guest).context(format!(
+                        "Failed to create disassembler for contract '{}'",
+                        path
+                    ))?;
+
```
Comment (xermicus) [inline crates/solidity/src/zksolc/main.rs:212]: I think we could instruct the disassembler to emit gas usage (`disassembler.display_gas()?;`).

**Review by xermicus — COMMENTED** [review]
Thanks!

I think since we now have an assembly text format, [`fn build_assembly_text`](https://github.com/0xf333/revive/blob/0xf333/addeding_revivetool_features/crates/llvm-context/src/polkavm/mod.rs#L23) should be fixed too

File: crates/solidity/src/zksolc/main.rs (hunk 3 above)
Comment (0xf333) [reply crates/solidity/src/zksolc/main.rs:208]: yep, I will add it now

File: crates/solidity/src/zksolc/main.rs (hunk 2 above)
Comment (0xf333) [reply crates/solidity/src/zksolc/main.rs:203]: ok

**Comment by 0xf333** [comment]
> Thanks!
> 
> I think since we now have an assembly text format, [`fn build_assembly_text`](https://github.com/0xf333/revive/blob/0xf333/addeding_revivetool_features/crates/llvm-context/src/polkavm/mod.rs#L23) should be fixed too

Yes, I actually planned `build_assembly_text` refactoring to go along with the `--zkasm` fix, which will be in my next pull request.
There seems to be more to it than simply giving it the correct assembly text.

File: crates/solidity/src/zksolc/main.rs [hunk 4]
```diff
@@ -190,18 +193,41 @@ fn main_inner() -> anyhow::Result<()> {
...
+                        ))
+                    })?;
+
+                let assembly_text = String::from_utf8(disassembled_code)
+                    .map_err(|error| anyhow::anyhow!(format!(
+                        "Failed to convert disassembled code to string for contract '{}'\n\nDue to:\n{}",
+                        path, error
+                    )))?;
+
+                println!("Contract `{}` assembly:\n\n{}", path, assembly_text);
```
Comment (xermicus) [inline crates/solidity/src/zksolc/main.rs:227]: Isn't this redudant now since we construct the disassembly in `crates/solidity/src/build/contract.rs`?

**Comment by xermicus** [comment]
> > Thanks!
> > I think since we now have an assembly text format, [`fn build_assembly_text`](https://github.com/0xf333/revive/blob/0xf333/addeding_revivetool_features/crates/llvm-context/src/polkavm/mod.rs#L23) should be fixed too
> 
> Yes, I actually planned `build_assembly_text` refactoring to go along with the `--zkasm` fix, which will be in my next pull request. There seems to be more to it than simply giving it the correct assembly text.

Right, so this will remove the redundant parts and just use `contract.build` everywhere

**Comment by 0xf333** [comment]
> Right, so this will remove the redundant parts and just use `contract.build` everywhere

Yeah, that's what I'm going for.

**Pull request merged**